### PR TITLE
fix: recursive call in connection `State` conversion to `i32`

### LINKED ISSUE #1009: Connection State conversion to i32 recursively calls itself causing stack overflow
## Bug Summary

The `from` implemented for `state` enum for conversion from `state` to `i32` in ics03-connection recursively calls itself causing an infinite loop and eventually overflowing the stack. 
For ref: https://github.com/cosmos/ibc-rs/blob/3894b50929e207eea3653e8bea3c70048d783d8c/ibc-core/ics03-connection/types/src/connection.rs#L560-L564

## Details

To reproduce the issue, try the following.
```rs
let state = ibc_core_connection_types::State::INIT;
let state_in_i32 = i32::from(state); // Would cause stack overflow here
```

## Version

0.48.1

### PR DESCRIPTION
**Pull request by dhruvja**
Closes: #1009 

## Description

Since the conversion from `state` to `i32` causes a recursive call, it can be solved by matching the enum variants and returning the i32 for the respective variant.

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [x] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: ibc-core/ics03-connection/types/src/connection.rs [hunk 1]
```diff
@@ -559,6 +559,11 @@ impl TryFrom<i32> for State {
 
 impl From<State> for i32 {
     fn from(value: State) -> Self {
-        value.into()
+        match value {
+            State::Uninitialized => 0,
+            State::Init => 1,
+            State::TryOpen => 2,
+            State::Open => 3,
+        }
```
Comment (mina86) [inline ibc-core/ics03-connection/types/src/connection.rs]:
```suggestion
        value as i32
```

**Review by mina86 — COMMENTED** [review]
Also, while you’re doing this could you replace `TryFrom<i32>` implementation to call `Self::from_i32` (or the other way around)?

**Review by Farhad-Shabani — APPROVED** [review]
Nice catch @dhruvja, thank you for the fix.

**Comment by dhruvja** [comment]
> Nice catch @dhruvja, thank you for the fix.

Spent quite a while figuring out the vague `stack overflow` bug lol.

**Pull request merged**